app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/PrivacyOptionsView.java

```diff
@@ -87,20 +87,31 @@ protected void updateUI() {
         permissionsTitleText.setText(getContext().getString(R.string.security_options_permissions_title, getContext().getString(R.string.app_name)));
 
         mPermissionButtons = new ArrayList<>();
-        mPermissionButtons.add(Pair.create(findViewById(R.id.cameraPermissionSwitch), Manifest.permission.CAMERA));
         mPermissionButtons.add(Pair.create(findViewById(R.id.microphonePermissionSwitch), Manifest.permission.RECORD_AUDIO));
-        mPermissionButtons.add(Pair.create(findViewById(R.id.locationPermissionSwitch), Manifest.permission.ACCESS_FINE_LOCATION));
         mPermissionButtons.add(Pair.create(findViewById(R.id.storagePermissionSwitch), Manifest.permission.WRITE_EXTERNAL_STORAGE));
 
         if (DeviceType.isOculusBuild() || DeviceType.isWaveBuild() || DeviceType.isPicoVR()) {
             findViewById(R.id.cameraPermissionSwitch).setVisibility(View.GONE);
+        } else {
+            mPermissionButtons.add(Pair.create(findViewById(R.id.cameraPermissionSwitch), Manifest.permission.CAMERA));
         }
+
         if (DeviceType.isOculusBuild()) {
             findViewById(R.id.locationPermissionSwitch).setVisibility(View.GONE);
+        } else {
+            mPermissionButtons.add(Pair.create(findViewById(R.id.locationPermissionSwitch), Manifest.permission.ACCESS_FINE_LOCATION));
         }
 
-        for (Pair<SwitchSetting, String> button: mPermissionButtons) {
-            button.first.setChecked(mWidgetManager.isPermissionGranted(button.second));
+        for (Pair<SwitchSetting, String> button : mPermissionButtons) {
+            if (button.second.equals(Manifest.permission.ACCESS_FINE_LOCATION)) {
+                boolean hasCoarseLocation = mWidgetManager.isPermissionGranted(Manifest.permission.ACCESS_COARSE_LOCATION);
+                boolean hasFineLocation = mWidgetManager.isPermissionGranted(Manifest.permission.ACCESS_FINE_LOCATION);
+                button.first.setChecked(hasFineLocation);
+                View locationWarning = findViewById(R.id.locationPermissionWarning);
+                locationWarning.setVisibility((hasCoarseLocation && !hasFineLocation) ? VISIBLE : GONE);
+            } else {
+                button.first.setChecked(mWidgetManager.isPermissionGranted(button.second));
+            }
```

Comment:
So this is adding a lot of code for the most uncommon case, to the point that the most common case is "lost" in an else block. 

I do have a suggestion, WDYT about doing all this outside the `for` loop? You can even extract it to a private method `showFineLocationWarningIfNeeded()` or something like that.

Reply:
I simplified things by using the bindings instead of `findViewById`, and by reordering a bit the code.